File: src/com/minhle/practiceleetcode/OddEvenLinkedList.java
package com.minhle.practiceleetcode;

public class OddEvenLinkedList {
    public ListNode oddEvenList(ListNode head) {
        ListNode curr = head;

        boolean isOdd = true;

        ListNode odds = new ListNode();
        ListNode currOdds = odds;

        ListNode evens = new ListNode();
        ListNode currEvens = evens;

        while (curr != null) {
            if (isOdd) {
                currOdds.next = new ListNode(curr.val);
                currOdds = currOdds.next;
            }
            else {
                currEvens.next = new ListNode(curr.val);
                currEvens = currEvens.next;
            }
            curr = curr.next;
            isOdd = !isOdd;
        }

        odds = odds.next;
        evens = evens.next;
        currOdds.next = evens;

        return odds;
    }
}

File: src/com/minhle/practiceleetcode/OddEvenLinkedList.cpp
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if (head == nullptr || head->next == nullptr)
        {
            return head;
        }
        auto *oddHead = head;
        auto *evenHead = head->next;
        auto *odd = oddHead;
        auto *even = evenHead;
        while (odd != nullptr && even != nullptr && odd->next != nullptr && even->next != nullptr)
        {
            odd->next = odd->next->next;
            odd = odd->next;
            even->next = even->next->next;
            even = even->next;
        }
        odd->next = evenHead;
        return oddHead;
    }
};
